idea/src/org/jetbrains/kotlin/idea/quickfix/MakeTypeParameterReifiedAndFunctionInlineFix.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2010-2017 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jetbrains.kotlin.idea.quickfix
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import org.jetbrains.kotlin.diagnostics.Diagnostic
+import org.jetbrains.kotlin.diagnostics.Errors
+import org.jetbrains.kotlin.idea.core.KotlinNameSuggester
+import org.jetbrains.kotlin.idea.core.addTypeParameter
+import org.jetbrains.kotlin.idea.search.usagesSearch.descriptor
+import org.jetbrains.kotlin.lexer.KtTokens
+import org.jetbrains.kotlin.psi.*
+import org.jetbrains.kotlin.psi.psiUtil.getStrictParentOfType
+
+class MakeTypeParameterReifiedAndFunctionInlineFix(
+        typeReference: KtTypeReference,
+        private val function: KtNamedFunction,
+        private val typeParameter: KtTypeParameter?
+) : KotlinQuickFixAction<KtTypeReference>(typeReference) {
+
+    private val inlineFix = AddInlineToFunctionWithReifiedFix(function)
+
+    override fun getText() = "Make type parameter reified and function inline"
+
+    override fun getFamilyName() = text
+
+    override fun invoke(project: Project, editor: Editor?, file: KtFile) {
+        val element = element ?: return
+
+        if (typeParameter != null) {
+            inlineFix.invoke(project, editor, file)
+            AddModifierFix(typeParameter, KtTokens.REIFIED_KEYWORD).invoke(project, editor, file)
+        }
+        else {
+            inlineFix.invoke(project, editor, file)
+            val psiFactory = KtPsiFactory(project)
+            val typeParameterName = KotlinNameSuggester.suggestNameByName("T") { name ->
+                function.typeParameters.none { it.name == name }
+            }
+            function.addTypeParameter(psiFactory.createTypeParameter("reified $typeParameterName"))
+            element.replace(psiFactory.createExpression(typeParameterName))
```

Comment:
All this else branch together with examples looks very dangerous to me. Looks like these actions change function semantics so I'd not recommend to perform them. I'd prefer to use this quick-fix only if type parameter exists.

Reply:
Fixed